ntoskrnl/rtl/misc.c

```diff
@@ -62,6 +63,140 @@ RtlGetVersion(IN OUT PRTL_OSVERSIONINFOW lpVersionInformation)
+{
+    HANDLE Key;
+    BOOLEAN Success;
+    ULONG ReturnedLength;
+    OBJECT_ATTRIBUTES ObjectAttributes;
+    NTSTATUS Status;
+    PKEY_VALUE_PARTIAL_INFORMATION BufferKey;
+    ULONG BufferKeyLength = sizeof(PKEY_VALUE_PARTIAL_INFORMATION) + (256 * sizeof(WCHAR));
+    UNICODE_STRING NtProductType;
+    static UNICODE_STRING WorkstationProduct = RTL_CONSTANT_STRING(L"WinNT");
+    static UNICODE_STRING LanManProduct = RTL_CONSTANT_STRING(L"LanmanNT");
+    static UNICODE_STRING ServerProduct = RTL_CONSTANT_STRING(L"ServerNT");
+    static UNICODE_STRING KeyProduct = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\SYSTEM\\CurrentControlSet\\Control\\ProductOptions");
+    static UNICODE_STRING ValueNtProduct = RTL_CONSTANT_STRING(L"ProductType");
+    PAGED_CODE();
+
+    /* Before doing anything else we must allocate some buffer space in the pool */
+    BufferKey = ExAllocatePoolWithTag(PagedPool, BufferKeyLength, PRODUCT_TAG);
+    if (!BufferKey)
+    {
+        /* We failed to allocate pool memory, bail out */
+        DPRINT1("RtlGetNtProductType(): Memory pool allocation has failed!\n");
+        Success = FALSE;
+        goto Exit;
+    }
+
+    /* Initialize the object attributes to open our key */
+    InitializeObjectAttributes(&ObjectAttributes,
+                               &KeyProduct,
+                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+                               NULL,
+                               NULL);
+
+    /* Open the key */
+    Status = ZwOpenKey(&Key, KEY_QUERY_VALUE, &ObjectAttributes);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlGetNtProductType(): The ZwOpenKey() function has failed! (Status: 0x%lx)\n", Status);
+        Success = FALSE;
+        goto Exit;
+    }
+
+    /* Now it's time to query the value from the key to check what kind of product the OS is */
+    Status = ZwQueryValueKey(Key,
+                             &ValueNtProduct,
+                             KeyValuePartialInformation,
+                             BufferKey,
+                             BufferKeyLength,
+                             &ReturnedLength);
+
+    /* Free the key from the memory */
+    ZwClose(Key);
+
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlGetNtProductType(): The ZwQueryValueKey() function has failed! (Status: 0x%lx)\n", Status);
+        Success = FALSE;
+        goto Exit;
+    }
+
+    /*
+     * Do a sanity check before we get the product type of the operating system
+     * so that we're sure the type of the value we've got is actually correct.
+    */
+    if (BufferKey->Type != REG_SZ)
+    {
+        /* We've got something else, so bail out */
+        DPRINT1("RtlGetNtProductType(): An invalid value type has been found!\n");
+        Success = FALSE;
+        goto Exit;
+    }
+
+    /* Initialize the Unicode string with the data from the registry value */
+    RtlInitUnicodeString(&NtProductType, (PWCHAR)BufferKey->Data);
```

Comment:
You can't assume this is null terminated.
It is probably better to use `DataLength` (which _can_ include the null terminator)